domain/externalcontroller/state/state.go

```diff
@@ -127,57 +128,131 @@ func (st *State) UpdateExternalController(
 		return errors.Trace(err)
 	}
 
-	cID := ci.ControllerTag.Id()
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		return st.updateExternalControllerTx(ctx, ci, modelUUIDs, tx)
+	})
+
+	return errors.Trace(err)
+}
+
+// ImportExternalControllers imports the list of MigrationControllerInfo
+// external controllers on one single transaction.
+func (st *State) ImportExternalControllers(ctx context.Context, infos []externalcontroller.MigrationControllerInfo) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
 
 	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
-		q := `
+		for _, ci := range infos {
+			err := st.updateExternalControllerTx(
+				ctx,
+				crossmodel.ControllerInfo{
+					ControllerTag: ci.ControllerTag,
+					Addrs:         ci.Addrs,
+					Alias:         ci.Alias,
+					CACert:        ci.CACert,
+				},
+				ci.ModelUUIDs,
+				tx,
+			)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
+}
+
+func (st *State) updateExternalControllerTx(
+	ctx context.Context,
+	ci crossmodel.ControllerInfo,
+	modelUUIDs []string,
+	tx *sql.Tx,
+) error {
+	cID := ci.ControllerTag.Id()
+	q := `
 INSERT INTO external_controller (uuid, alias, ca_cert)
 VALUES (?, ?, ?)
   ON CONFLICT(uuid) DO UPDATE SET alias=excluded.alias, ca_cert=excluded.ca_cert`[1:]
 
-		if _, err := tx.ExecContext(ctx, q, cID, ci.Alias, ci.CACert); err != nil {
-			return errors.Trace(err)
-		}
+	if _, err := tx.ExecContext(ctx, q, cID, ci.Alias, ci.CACert); err != nil {
+		return errors.Trace(err)
+	}
 
-		addrsBinds, addrsAnyVals := database.SliceToPlaceholder(ci.Addrs)
-		q = fmt.Sprintf(`
+	addrsBinds, addrsAnyVals := database.SliceToPlaceholder(ci.Addrs)
+	q = fmt.Sprintf(`
 DELETE FROM external_controller_address
```

Comment:
Do we currently update controller addresses when importing external controllers? 

I know it's unlikely, but what if the migration target has a more up-to-date representation of the controller than the encoded model?

Reply:
currently (mongo state) we also upsert during migration (`upsertExternalControllerOp()` is ultimately called by `(ImportExternalControllers) Execute()`: https://github.com/juju/juju/blob/main/state/externalcontroller.go#L293:L295